IdentityCore/src/cache/mac/MSIDMacAppCredentialCacheItem.m

```diff
@@ -0,0 +1,194 @@
+        MSIDCredentialCacheItem *appToken = [cacheObjects objectForKey:appTokenKey];
+        NSDictionary *atDict = [appToken jsonDictionary];
+        
+        if (atDict)
+        {
+            [dictionary setObject:atDict forKey:key];
+        }
+    }
+    
+    return dictionary;
+}
+
+- (void)setAppCredential:(MSIDCredentialCacheItem *)token forKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects setObject:token forKey:key];
+    });
+}
+
+-(MSIDCredentialCacheItem *)appCredentialForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    __block MSIDCredentialCacheItem *appCredential = nil;
+    
+    dispatch_sync(self.queue, ^{
+        appCredential = [self.cacheObjects objectForKey:key];
+    });
+    
+    return appCredential;
+}
+
+- (void)removeAppCredentialForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects removeObjectForKey:key];
+    });
+}
+
+- (void)mergeAppCredential:(MSIDMacAppCredentialCacheItem *)credential
+{
+    NSDictionary *copy = [credential getCopy];
+    
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects addEntriesFromDictionary:copy];
+    });
+}
+
+- (NSArray<MSIDCredentialCacheItem *> *)appCredentialsWithKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    if (key.account && key.service)
+    {
+        MSIDCredentialCacheItem *credential = [self appCredentialForKey:key];
+        if (credential)
+        {
+            return @[credential];
+        }
+        
+        return nil;
+    }
+    
+    NSMutableArray *subPredicates = [[NSMutableArray alloc] init];
+    
+    if (key.clientId)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.clientId == %@", key.clientId]];
+    if (key.environment)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.environment == %@", key.environment]];
+    if (key.homeAccountId)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.homeAccountId == %@", key.homeAccountId]];
+    if (key.credentialType)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.credentialType == %d", key.credentialType]];
+    if (key.target)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.target == %@", key.target]];
+    if (key.realm)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.realm == %@", key.realm]];
+    
+    // Combine all sub-predicates with AND:
+    NSPredicate *matchAttributes = [NSCompoundPredicate andPredicateWithSubpredicates:subPredicates];
+    
+    NSDictionary *cacheObjects = [self getCopy];
+    return [[cacheObjects allValues] filteredArrayUsingPredicate:matchAttributes];;
+}
+
+- (NSDictionary *)getCopy
+{
+    __block NSMutableDictionary *copy;
+
+    dispatch_sync(self.queue, ^{
+        copy = [self.cacheObjects mutableDeepCopy];
```

Comment:
hmmm, this seems a bit expensive. Can we avoid making a copy every time?

Reply:
Copy might be more expensive than lock-and-search :P